decoders/a7105/pd.py

```diff
@@ -0,0 +1,356 @@
+        '''Returns the collected MOSI bytes of a multi byte command.'''
+        return [b[0] for b in self.mb]
+
+    def miso_bytes(self):
+        '''Returns the collected MISO bytes of a multi byte command.'''
+        return [b[1] for b in self.mb]
+
+    def decode_command(self, pos, b):
+        '''Decodes the command byte 'b' at position 'pos' and prepares
+        the decoding of the following data bytes.'''
+        c = self.parse_command(b)
+        if c is None:
+            self.warn(pos, 'unknown command')
+            return
+
+        self.cmd, self.dat, self.min, self.max = c
+
+        if self.cmd in ('W_REGISTER', 'R_REGISTER'):
+            # Don't output anything now, the command is merged with
+            # the data bytes following it.
+            self.mb_s = pos[0]
+        else:
+            self.putp(pos, self.ann_cmd, self.format_command())
+
+    def format_command(self):
+        '''Returns the label for the current command.'''
+        return 'Cmd {}'.format(self.cmd)
+
+    def parse_command(self, b):
+        '''Parses the command byte.
+
+        Returns a tuple consisting of:
+        - the name of the command
+        - additional data needed to dissect the following bytes
+        - minimum number of following bytes
+        - maximum number of following bytes
+        '''
+        
+        if b == 0x05:
+            return ('W_TX_FIFO', None, 1, 32)
+        elif b == 0x45:
+            return ('R_RX_FIFO', None, 1, 32)            
+        if b == 0x06:
+            return ('W_ID', None, 1, 4)
+        elif b == 0x46:
+            return ('R_ID', None, 1, 4)
+        elif (b & 0b10000000) == 0:
+            if (b & 0b01000000) == 0:
+                c = 'W_REGISTER'
+            else:
+                c = 'R_REGISTER'
+            d = b & 0b00111111
+            return (c, d, 1, 1)
+		
+        else:
+            cmd = b & 0b11110000        
+            if cmd == 0b10000000:
+                return ('SLEEP_MODE', None, 0, 0)
+            if cmd == 0b10010000:
+                return ('IDLE_MODE', None, 0, 0)
+            if cmd == 0b10100000:
+                return ('STANDBY_MODE', None, 0, 0)
+            if cmd == 0b10110000:
+                return ('PLL_MODE', None, 0, 0)
+            if cmd == 0b11000000:
+                return ('RX_MODE', None, 0, 0)
+            if cmd == 0b11010000:
+                return ('TX_MODE', None, 0, 0)
+            if cmd == 0b11100000:
+                return ('FIFO_WRITE_PTR_RESET', None, 0, 0)
+            if cmd == 0b11110000:
+                return ('FIFO_READ_PTR_RESET', None, 0, 0)
+
+    def decode_register(self, pos, ann, regid, data):
+        '''Decodes a register.
+
+        pos   -- start and end sample numbers of the register
+        ann   -- is the annotation number that is used to output the register.
+        regid -- may be either an integer used as a key for the 'regs'
+                 dictionary, or a string directly containing a register name.'
```

Comment:
does this decoder actually use both forms? or is this left over code from some development work?

Reply:
I code this decoder when I need to analyze the MESSAGE to A7105 when I work out the WFLY2 protocol this project https://github.com/pascallanger/DIY-Multiprotocol-TX-Module, it just works and decode the messages properly.